geary-papermc-tracking/src/main/kotlin/com/mineinabyss/geary/papermc/tracking/blocks/BlockTracking.kt

```diff
@@ -0,0 +1,95 @@
+import org.bukkit.block.data.type.Tripwire
+
+val gearyBlocks by DI.observe<BlockTracking>()
+interface BlockTracking {
+    val blockMap: Map<BlockData, Int>
+    val blockProvider: GearyBlockProvider
+    val block2Prefab: Block2Prefab
+    val blockPrefabs: GearyBlockPrefabQuery
+
+    fun createBlockData(prefabKey: PrefabKey): BlockData? = blockProvider.serializePrefabToBlockData(prefabKey)
+
+    companion object : GearyAddonWithDefault<BlockTracking> {
+        override fun default(): BlockTracking = object : BlockTracking {
+            override val blockMap = createBlockMap()
+            override val blockProvider = GearyBlockProvider()
+            override val blockPrefabs = GearyBlockPrefabQuery()
+            override val block2Prefab = Block2Prefab()
+        }
+
+        override fun BlockTracking.install() {
+            geary.pipeline.addSystems(
+                TrackOnSetBlockComponent()
+            )
+        }
+
+        private fun createBlockMap(): MutableMap<BlockData, Int> {
+            return mutableMapOf<BlockData, Int>().apply {
+                // Calculates tripwire states
+                for (i in 0..127) {
+                    val tripWireData = Material.TRIPWIRE.createBlockData() as Tripwire
+                    if (i and 1 == 1) tripWireData.setFace(BlockFace.NORTH, true)
+                    if (i shr 1 and 1 == 1) tripWireData.setFace(BlockFace.EAST, true)
+                    if (i shr 2 and 1 == 1) tripWireData.setFace(BlockFace.SOUTH, true)
+                    if (i shr 3 and 1 == 1) tripWireData.setFace(BlockFace.WEST, true)
+                    if (i shr 4 and 1 == 1) tripWireData.isPowered = true
+                    if (i shr 5 and 1 == 1) tripWireData.isDisarmed = true
+                    if (i shr 6 and 1 == 1) tripWireData.isAttached = true
+
+                    putIfAbsent(tripWireData, i)
+                }
+
+                // Calculates noteblock states
+                // We do 25-825 to skip PIANO at first
+                for (j in 50..799) {
+                    //val id = if (blocky.config.noteBlocks.restoreNormalFunctionality && j <= 50) j + 799 else j
+                    val noteBlockData = Material.NOTE_BLOCK.createBlockData() as NoteBlock
+                    noteBlockData.instrument = Instrument.getByType((j / 50 % 400).toByte()) ?: continue
+
+                    noteBlockData.note = Note((j % 25))
+                    noteBlockData.isPowered = j / 25 % 2 == 1
+
+                    putIfAbsent(noteBlockData, j - 49)
+                }
+                // Restore functionality bs
+                for (j in 1..49) {
+                    val noteBlockData = Material.NOTE_BLOCK.createBlockData() as NoteBlock
+                    noteBlockData.instrument = Instrument.PIANO
+                    noteBlockData.note = Note((j % 25))
+                    noteBlockData.isPowered = j / 25 % 2 == 1
+
+                    putIfAbsent(noteBlockData, j + 750)
+                }
```

Comment:
NoteBlock-type BlockData calculation. Split in two due to original concept of restoring redstone functionality
Might finish this at a later date but this causes no issues in the current impl